platform/broadcom/sonic-platform-modules-delta/ag5648/modules/dni_emc2305.c

```diff
@@ -0,0 +1,390 @@
+  mutex_unlock(&data->lock);
+  return sprintf(buf, "%d\n", val);
+}
+
+static ssize_t set_pwm(struct device *dev, struct device_attribute *devattr,
+                       const char *buf, size_t count)
+{
+  struct sensor_device_attribute *attr = to_sensor_dev_attr(devattr);
+  struct i2c_client *client = to_i2c_client(dev);
+  struct emc2305_data *data = i2c_get_clientdata(client);
+  unsigned long val;
+  int ret;
+
+  ret = kstrtoul(buf, 10, &val);
+
+  if (ret)
+  {
+    return ret;
+  }
+  if (val > 255)
+  {
+    return -EINVAL;
+  }
+
+  mutex_lock(&data->lock);
+  i2c_smbus_write_byte_data(client,
+                            EMC2305_REG_FAN_DRIVE(attr->index),
+                            val);
+  mutex_unlock(&data->lock);
+
+  return count;
+}
+
+static int emc2305_detect(struct i2c_client *client,
+                          struct i2c_board_info *info)
+{
+  struct i2c_adapter *adapter = client->adapter;
+  int vendor, device;
+
+  if (!i2c_check_functionality(adapter, I2C_FUNC_SMBUS_BYTE_DATA |
+                               I2C_FUNC_SMBUS_WORD_DATA))
+  {
+    return -ENODEV;
+  }
+
+  vendor = i2c_smbus_read_byte_data(client, EMC2305_REG_VENDOR);
+  if (vendor != EMC2305_VENDOR)
+  {
+    return -ENODEV;
+  }
+
+  device = i2c_smbus_read_byte_data(client, EMC2305_REG_DEVICE);
+  if (device != EMC2305_DEVICE)
+  {
+    return -ENODEV;
+  }
+
+  strlcpy(info->type, "emc2305", I2C_NAME_SIZE);
+
+  return 0;
+}
+
+static int emc2305_probe(struct i2c_client *client,
+                         const struct i2c_device_id *id)
+{
+  struct emc2305_data *data;
+  int err;
+  int i;
+
+  data = devm_kzalloc(&client->dev, sizeof(struct emc2305_data),
+                      GFP_KERNEL);
+  if (!data)
+  {
+    return -ENOMEM;
+  }
+
+  i2c_set_clientdata(client, data);
+  mutex_init(&data->lock);
+
+  dev_info(&client->dev, "%s chip found\n", client->name);
+
+  data->attrs.attrs = emc2305_attr;
+  err = sysfs_create_group(&client->dev.kobj, &data->attrs);
+  if (err)
+  {
+    return err;
+  }
+
+  data->hwmon_dev = hwmon_device_register(&client->dev);
+  if (IS_ERR(data->hwmon_dev))
+  {
+    err = PTR_ERR(data->hwmon_dev);
+    goto exit_remove;
+  }
+
+  for (i = 0; i < 4; i++)
+  {
+    /* set minimum drive to 0% */
+    i2c_smbus_write_byte_data(client, EMC2305_REG_FAN_MIN_DRIVE(i), FAN_MINIMUN);
+    i2c_smbus_write_byte_data(client, EMC2305_REG_FAN_CONF(i), FAN_RPM_BASED);
+  }
+
+  return 0;
+
+exit_remove:
+  sysfs_remove_group(&client->dev.kobj, &data->attrs);
+  return err;
+}
+
+static int emc2305_remove(struct i2c_client *client)
+{
+  struct emc2305_data *data = i2c_get_clientdata(client);
+
+  hwmon_device_unregister(data->hwmon_dev);
+  sysfs_remove_group(&client->dev.kobj, &data->attrs);
+  return 0;
+}
+
+module_i2c_driver(emc2305_driver);
+
+MODULE_AUTHOR("Neal Tai<neal.tai@deltaww.com>");
+MODULE_DESCRIPTION("SMSC EMC2305 fan controller driver");
+MODULE_LICENSE("GPL");
```

Comment:
what is the difference between this one and the one for 9032 platform. i saw one there too.

Reply:
There are four fan trays on ag5648 & five on ag9032v1.

Reply:
in terms of the driver, they should be the same. can we re-use the existing driver?

Reply:
OK! I will use the existing driver.